src/main/java/com/amazon/opendistroforelasticsearch/security/auditlog/impl/AuditCategory.java

```diff
@@ -0,0 +1,44 @@
+package com.amazon.opendistroforelasticsearch.security.auditlog.impl;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Objects;
+
+public enum AuditCategory {
+    BAD_HEADERS,
+    FAILED_LOGIN,
+    MISSING_PRIVILEGES,
+    GRANTED_PRIVILEGES,
+    OPENDISTRO_SECURITY_INDEX_ATTEMPT,
+    SSL_EXCEPTION,
+    AUTHENTICATED,
+    COMPLIANCE_DOC_READ,
+    COMPLIANCE_DOC_WRITE,
+    COMPLIANCE_EXTERNAL_CONFIG,
+    COMPLIANCE_INTERNAL_CONFIG_READ,
+    COMPLIANCE_INTERNAL_CONFIG_WRITE;
+
+    private static Logger log = LogManager.getLogger(AuditCategory.class);
+
+    public static EnumSet<AuditCategory> parse(final List<String> categories) {
+        EnumSet<AuditCategory> set = EnumSet.noneOf(AuditCategory.class);
+        if (categories == null)
+            return set;
+
+        categories
+                .stream()
+                .filter(Objects::nonNull)
+                .map(category -> category.toUpperCase())
+                .forEach(category -> {
+                    try {
+                        set.add(AuditCategory.valueOf(category));
+                    } catch (IllegalArgumentException ex) {
+                        log.error("Invalid category type: {} Exception: {}", category, ex.getMessage());
```

Comment:
I am not sure that it is sufficient to log an error message and continue. It is possible that a category has a typo that causes parsing exception and it may go unnoticed (lost in log files).

Reply:
I believe if we can recover from some error, then complain and recover. I hope error logs are not that casually ignored.  
In this case, if incorrect category is provided, simply complain and ignore it.

Reply:
Imagine that the `AuditCategory` is constructed using a call to REST API from `curl` and somebody mistypes a category name. The return code of REST API would be 200 and the user who called REST API would assume that it succeeded and even not bother to look at the error log.

Reply:
That should directly return a bad request if you are trying to make a log message for an incorrect category type. 
We can try to convert to Enum and then throw error in the validator of that rest api handler.  Not sure if we should handle it here.

Reply:
In the example, I refer to constructing `EnumSet<AuditCategory>` from a string passed as REST API parameter, not to constructing `AuditMessage` with an invalid `AuditCategory`.

Reply:
Alright. Do you want to throw exception here instead ? Or 
Use something like:

```
        return categories
                .stream()
                .filter(Objects::nonNull)
                .map(String::toUpperCase)
                .map(AuditCategory::valueOf)
                .collect(Collectors.toCollection(() -> set));
```
And let java throw the IllegalArgumentException ?

Reply:
This way, it will throw exception and prevent the service from starting up if elasticsearch.yml is configured incorrectly and can also assert for exceptions in REST api validators and throw bad request appropriately.

Reply:
It is OK to rely on `valueOf()` to throw an `IllegalArgumentException` or to convert it to a checked exception. I am slightly more inclined to `IllegalArgumentException`.

Reply:
AuditCategory::valueOf will throw IllegalArgumentException if cannot be converted to enum. Am I missing something?

Reply:
@vrozov The current behavior is to actually post error and ignore invalid enum and have some tests to do so.

Reply:
> AuditCategory::valueOf will throw IllegalArgumentException if cannot be converted to enum. Am I missing something?

No. I was referring to the same.

> The current behavior is to actually post error and ignore invalid enum and have some tests to do so

IMO, there is a problem with an approach that tries to correct an invalid input. At a minimum, `AuditCategory` should allow a caller to get the list of categories that were not parsed or throw an exception.

Reply:
Can we have this when we need it?

Reply:
What is your concern? Changing tests?

Reply:
I meant to say, lets do it when we have a use-case ready. For sure, this will come up in future and let's target it that time. Maybe we will have more requirements.

Reply:
Could you explain the rationale/use-case for including this now? I think we are not on the same page. Where and what do you want do with string values that cannot be parsed ?

Reply:
The current behavior is to ignore an invalid value and continue. The invalid value may come from a configuration file (the only way supported now) or from an API call with a request to update/change configuration (in the future). In both cases, likely a user (system admin) made a mistake and simply mistyped **valid**  audit category name that the user (system admin) wants to be in effect. How quickly the user (system admin) may realize that the system ignored or partially ignored the input is questionable.

Reply:
Note that the original code provided a better hint in the logging message of what setting has an unknown/invalid category. It may be good to keep it the same way should you decide to keep the existing behavior.  It is not necessary to log the `InvalidArgumentException` message.

Reply:
I will revert to throwing exception and will update tests to check for exception.